add knowledge center mysql generated column article (#904)
ariga/atlas: doc/md/knowledge-center/mysql/generated_columns.md

---
id: mysql-generated-columns

> title: How to Use Generated Columns in MySQL with Atlas

Reviewer: Generated columns are part of the SQL standard. i.e. support by ~all databases.

Can we make this doc more general and map the differences between dialects in one section? Another option is to use tabs in case grammar or examples are different between dialects (it's mostly the same). See examples in https://atlasgo.io/ddl/sql#schema.

Reviewer 2: Hey , as we spoke on the phone this is on purpose to increase search visibility, we will have a specific article per DB

Reviewer: OK, I understand from @rotemtam that it is going to be copied per dialect. My request please is to move it to a different page. Something like [entgo.io/tutorials](https://entgo.io/docs/tutorial-setup). 

<img width="660" alt="image" src="https://user-images.githubusercontent.com/7413593/177094954-1b185acd-47da-4a4d-ad1c-9c806a07caed.png">

Author: What is the decision on this?

Reviewer 2: My opinion is that we should go ahead like this and create a dedicated section once we have more content. @a8m ?

Reviewer: Let’s put it in a different category as it’s not part of the documentation and we prefer to start clean without tech debt. I plan to place the checks/lint rules in a separate category as well.

Please, let me know if you need help with this.

Author: I have moved the knowledge center to its own categorical page. My only concern is that it's called a Knowledge Center/Base but the URL is `/knowledge`. Rotem has moved the task of the page's stylization to a new pulll request.

slug: /knowledge/mysql/generated-columns
---

> **MySQL** is a popular open-source relational database that lets you store data. **Generated Columns** are a feature of MySQL that allows you to store and read data that is dependent on other data; without requiring complex INSERT or UPDATE queries.

Reviewer 3: ```suggestion
**MySQL** is a popular open-source relational database that lets you store data. **Generated columns** are a feature of MySQL that allows you to store and read data that is dependent on other data; without requiring complex INSERT or UPDATE queries.
```

> **MySQL** is a popular open-source relational database that lets you store data. **Generated Columns** are a feature of MySQL that allows you to store and read data that is dependent on other data; without requiring complex INSERT or UPDATE queries.

Reviewer 2: ```suggestion
**MySQL** is a popular open-source relational database. **Generated Columns** are a feature of MySQL that allows you to define tables with columns whose value is a function of the value stored in other columns; without requiring complex expressions in SELECT, INSERT or UPDATE queries.
```

> **MySQL** is a popular open-source relational database. **Generated columns** are a feature of MySQL that allows you to define tables with columns whose value is a function of the value stored in other columns; without requiring complex expressions in SELECT, INSERT or UPDATE queries.

Reviewer: ```suggestion
**MySQL** is a popular open-source relational database. **Generated columns** are a feature of MySQL that allows you to define tables with columns whose value is a function of the value stored in other columns; without requiring complex expressions in `SELECT`, `INSERT` or `UPDATE` queries.
```

> **MySQL** is a popular open-source relational database. **Generated columns** are a feature of MySQL that allows you to define tables with columns whose value is a function of the value stored in other columns; without requiring complex expressions in SELECT, INSERT or UPDATE queries.

Reviewer: ```suggestion
**MySQL** is a popular open-source relational database. **Generated columns** are a feature of MySQL that allows you to define tables with columns whose value is a function of the value stored in other columns; without requiring complex expressions in SELECT, INSERT or UPDATE queries.
```

This sentence is not 100% correct. A definition like this is also valid:

```
create table t(c int as (1+f()));
```

## What are Generated Columns?

> Generated Columns are columns that contain values calculated by expressions which can be dependent on other columns; in a similar manner to formulas in a spreadsheet. There are two types of Generated Columns in MySQL: _Stored_ and _Virtual_.

Reviewer 3: ```suggestion
Generated columns are columns that contain values calculated by expressions which can be dependent on other columns; in a similar manner to formulas in a spreadsheet. There are two types of generated columns in MySQL: _Stored_ and _Virtual_.
```

### Stored Generated Columns

> Stored Generated Columns are stored and evaluated when a row is inserted or updated. As a result, stored generated columns use disk space in addition to CPU cycles during the execution of `INSERT` and `UPDATE` statements.

Reviewer 3: ```suggestion
Stored generated columns are stored and evaluated when a row is inserted or updated. As a result, stored generated columns use disk space in addition to CPU cycles during the execution of `INSERT` and `UPDATE` statements.
```

### Virtual Generated Columns

> Virtual Generated Columns are not stored, and only evaluated when a row is read _(after BEFORE [triggers](https://dev.mysql.com/doc/refman/5.7/en/trigger-syntax.html))_. As a result, virtual generated columns take no storage at the cost of CPU cycles for `SELECT` statements.

Reviewer 3: ```suggestion
Virtual generated columns are not stored, and only evaluated when a row is read _(after BEFORE [triggers](https://dev.mysql.com/doc/refman/5.7/en/trigger-syntax.html))_. As a result, virtual generated columns take no storage at the cost of CPU cycles for `SELECT` statements.
```

> Virtual generated columns are not stored, and only evaluated when a row is read _(after BEFORE [triggers](https://dev.mysql.com/doc/refman/5.7/en/trigger-syntax.html))_. As a result, virtual generated columns take no storage at the cost of CPU cycles for `SELECT` statements.

Reviewer: Maybe mention the pros/cons of each.

Author: In addition to the use-case ("When to Use Generated Columns"), these are the pros and cons.

### Limitations of Generated Columns

> Generated Column expressions must be deterministic which means that — given the same input — an expression must always produce the same output. As a result, Generated Columns can not be used with stored variables, functions, procedures, and subqueries; which could cause the output to be non-deterministic. Following this constraint, Generated Columns can not be used to generate random values. On the other hand, a Generated Column may reference other generated columns within the same table, as long as those columns are declared before the Generated Column. In contrast, a Generated Column may reference any non-generated column _regardless_ of its position within the table.

Reviewer 3: ```suggestion
Generated column expressions must be deterministic, meaning that — given the same input — an expression must always produce the same output. As a result, generated columns can not be used with stored variables, functions, procedures, and subqueries; which could cause the output to be non-deterministic. Following this constraint, Generated Columns can not be used to generate random values. On the other hand, a generated column may reference other generated columns within the same table, as long as those columns are declared before the generated column. In contrast, a generated column may reference any non-generated column _regardless_ of its position within the table.
```

> Generated column expressions must be deterministic which means that — given the same input — an expression must always produce the same output. As a result, generated columns can not be used with stored variables, functions, procedures, and subqueries; which could cause the output to be non-deterministic. Following this constraint, generated columns can not be used to generate random values. On the other hand, a generated column may reference any non-generated column _regardless_ of its position within the table row and any other generated column within the same table row, as long as those columns are declared before the generated column.

Reviewer: > and subqueries; 

Subqueries are never permitted.

Reviewer: I misread the sentence. Thanks

## When to use Generated Columns?

Generated columns should be used whenever you want to create a column with a value that can be directly determined from the values of other columns in the same row. In simpler words, for data that is dependent on other data. This saves the developer from complex application code that is prone to errors on `SELECT`, `INSERT` and `UPDATE` statements. It also ensures that data which must be consistent, stays consistent.

**MySQL Syntax for a Generated Column**

```sql
column_name data_type [GENERATED ALWAYS] AS (expr) [VIRTUAL | STORED] 
    [NOT NULL | NULL] [UNIQUE [KEY]] [[PRIMARY] KEY] [COMMENT 'string']
```

### Using Stored Generated Columns

> Stored Generated Columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored Generated Columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Reviewer 3: ```suggestion
Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.
```

> Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Author: That is for virtual columns and is already mentioned in the virtual column section. A virtual column can only be used with secondary indexes. Please see https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html#:~:text=A%20stored%20column%20does%20require%20storage%20space%20and%20can%20be%20indexed.

> Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Reviewer: When referring to indexes, it's usually secondary indexes (keys). The primary index is the primary key (index). Did you mean this? 

 > Stored generated columns should also be used when you want to index the column

When reading this, I got the impression that I should use "STORED" (and not "VIRTUAL") in case I want to index the column.

> Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Author: Yes. If you want to use a secondary index (which allows duplicates), you can use stored or virtual columns. However, if you want to use a primary index (no duplicates), then you must use a stored column.

> Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Reviewer: OK, can we change "index" to "primary key" in this sentence? As I mentioned above, it's common in the "industry" to call secondary indexes/keys just indexes, and primary indexes as primary keys.

> Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Author: An index isn't necessarily a primary key, but a primary key is an index.
- https://dev.mysql.com/doc/refman/8.0/en/create-index.html
- https://dev.mysql.com/doc/refman/8.0/en/mysql-indexes.html

I was told to prioritize correctness over simplicity with regards to these types of sentences, so by that objective the sentence should remain the same.

> Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Reviewer: From reading this sentence, I'll expect the following command to not pass:

<img width="649" alt="image" src="https://user-images.githubusercontent.com/7413593/177418240-aa5afb77-47ac-48c8-b9da-fbfa80418f3a.png">

As you can see, I created a **virtual** column and indexed it.

> Stored generated columns should be used for data _(in a table)_ that is read more frequently than it is updated. This saves CPU cycles while reading rows _(via `SELECT`)_. Stored generated columns should also be used when you want to index the column or use it as a foreign key constraint. Alternatively, use stored generated columns as a cache for complex conditions that are costly to calculate.

Author: The article later states:

> If you use the _InnoDB Storage Engine_, secondary indexes can be defined on virtual columns.

The focus of this comment is on the use-case for stored generated columns:

> Stored generated columns should also be used when you want to index the column

This doesn't imply that the following command will fail, but only that a virtual column is only valid with a secondary index, while a stored generated column can be used with any index.

Feel free to suggest a change to make this distinction clearer. However, specifying this difference in the stored generated column section would be wrong. For example, "Stored generated columns should also be used when you want to index the column _(except for secondary indexes)_" is an incorrect statement. You can use any index with a stored generated column, but only secondary indexes on the virtual column _(with the InnoDB Storage Engine)_.

#### Example

> The following example declares a Stored Generated Column in a table that stores the base and height of a triangle in the `base` and `height` column, then computes its area in `area` _(when triangles are inserted or updated)_.

Reviewer 3: ```suggestion
The following example declares a stored generated column in a table that stores the base and height of a triangle in the `base` and `height` column, then computes its area in `area` _(when triangles are inserted or updated)_.
```

```sql
CREATE TABLE triangles (
    base   DOUBLE,
    height DOUBLE,
    area   DOUBLE AS (base * height * 1/2) STORED
);
```

### Using Virtual Generated Columns

> Virtual Generated Columns should be used for computed data _(in a table)_ that is updated more frequently than it is read or computed data that is expensive to store on disk _(via `INSERT` or `UPDATE`)_. Since values are calculated on the fly, virtual generated columns are perfect for table columns that will have a new value for every `SELECT` statement. If you use the _InnoDB Storage Engine_, secondary indexes can be defined on virtual columns.

Reviewer 3: ```suggestion
Virtual generated columns should be used for computed data _(in a table)_ that is updated more frequently than it is read or computed data that is expensive to store on disk _(via `INSERT` or `UPDATE`)_. Since values are calculated on the fly, virtual generated columns are perfect for table columns that will have a new value for every `SELECT` statement. If you use the _InnoDB Storage Engine_, secondary indexes can be defined on virtual columns.
```

#### Example

> The following example declares a Virtual Generated Column in a table that stores the price and amount of products sold in the `price` and `quantity` column, then computes its `revenue` _(when products are read)_.

Reviewer 3: ```suggestion
The following example declares a virtual generated column in a table that stores the price and amount of products sold in the `price` and `quantity` column, then computes its `revenue` _(when products are read)_.
```

```sql
CREATE TABLE products (
    price    DOUBLE,
    quantity INT,
    revenue  DOUBLE AS (price * quantity) VIRTUAL
);
```

## Managing Generated Columns is easy with Atlas

> Managing Generated Columns and Database Schemas in MySQL is confusing and error-prone. [Atlas](https://atlasgo.io) is an open-source tool that allows you to manage your database using a simple declarative syntax (similar to Terraform). Instead of creating complex SQL statements that break upon schema migration, we will implement Generated Columns using Atlas.

Reviewer 3: ```suggestion
Managing generated columns and database schemas in MySQL is confusing and error-prone. [Atlas](https://atlasgo.io) is an open-source tool that allows you to manage your database using a simple declarative syntax (similar to Terraform). Instead of creating complex SQL statements that break upon schema migration, we will implement generated columns using Atlas.
```

### Getting started with Atlas

Install the latest version of Atlas using the [Guide to Setting Up Atlas](https://atlasgo.io/cli/getting-started/setting-up).

### Generated Column Syntax in Atlas

Use `as` in a column in a table to declare a MySQL generated column. For examples with other databases, read the [Atlas Generated Columns DDL](https://atlasgo.io/ddl/sql/#generated-columns).

```hcl
column "name" {
    type = data_type
    as {
        expr = expression
        type = [STORED | VIRTUAL]
    }
}
```

### Implementing Stored Generated Columns with Atlas

> The following example declares a Stored Generated Column in a table that stores the lengths of the sides of right-triangles in the `a` and `b` column, then computes the hypotenuse in `c` _(when triangles are inserted or updated)_.

Reviewer 3: ```suggestion
The following example declares a stored generated column in a table that stores the lengths of the sides of right-triangles in the `a` and `b` column, then computes the hypotenuse in `c` _(when triangles are inserted or updated)_.
```

        as  {
            expr  = "SQRT(a * a + b * b)"
            type  = STORED
        }
    }
}
```

Guarantee the table is created by applying the schema to the database.

```
atlas schema apply -u "mysql://root:pass@localhost:3306/example" -f atlas.hcl
```

Approve the schema migration plan that Atlas creates for you _(if applicable)_.

```
-- Planned Changes:
-- Create "triangles" table
CREATE TABLE "example"."triangles" ("a" numeric NOT NULL, "b" numeric NOT NULL, "hypotenuse" numeric NOT NULL GENERATED ALWAYS AS (SQRT(a * a + b * b)) STORED)
Use the arrow keys to navigate: ↓ ↑ → ←
? Are you sure?:
> Apply
Abort
```

Insert triangles into the table.

```sql
INSERT INTO triangles (a, b) VALUES (1,1);
INSERT INTO triangles (a, b) VALUES (3,4);
INSERT INTO triangles (a, b) VALUES (6,8);
```

Select all the triangles in the table using `SELECT  *  FROM triangles` to receive a table with the following output.

| a    | b    | c                  |
| :--- | :--- | :----------------- |
| 1    | 1    | 1.4142135623730951 |
| 3    | 4    | 5                  |
| 6    | 8    | 10                 |

### Implementing Virtual Generated Columns Columns with Atlas

> The following example declares a Virtual Generated Column in a TABLE that stores the first and last name of a person, and computes the full name of the person _(when people are selected)_.

Reviewer 3: ```suggestion
The following example declares a virtual generated column in a TABLE that stores the first and last name of a person, and computes the full name of the person _(when people are selected)_.
```
